doc/designs/migration.md

```diff
@@ -0,0 +1,456 @@
+to discover differences. We probably want to do only ADD and bail if we
+determine something will be a MOD or DEL.
+
+#### Excluding specific entries
+
+The ignore user/group options of migrate-ds were introduced so that non-compliant
+entries could be skipped to not block the migration. We can ignore these for now.
+
+#### Performance considerations
+
+Some benefits would be likely if common entry types were added in
+batches rather than individually. This would not easily allow for a
+"stop on failure" approach but could be considerably faster.
+
+The basic idea is that added entries of the same type (user, group, etc)
+would be accumulated and added together. The result would be iterated
+through and logged.
+
+The memberOf plugin should be disabled. A fixup task can be run post-migration
+to calculate the memberships.
+
+#### Idempotency
+
+The migrate-ds plugin manages idempotency because it skips over entries
+that already have been migrated (or exist). This new migration will handle
+it by merging the remote and local entries. This may not be desirable in
+all cases.
+
+#### Migration steps
+
+A simplistic view of the steps
+
+Set migration mode=True
+Disable compat
+Disable memberof
+Migrate schema
+Loop through the list of remote objects:
+  - fix up any DN values or other syntaxes
+  - remove conflicting attributes/objectclasses
+  - write new entry
+      - if entry already exists, merge them if possible preferring remote side
+Enable memberOf
+Create memberOf fixup task
+Enable compat
+Restart world
+Run ipa-server-upgrade
+
+### pure LDAP migration
+
+The current code generally works minus a few bugs and RFEs, some of which are resolved
+by doing an IPA-to-IPA migration instead. It should be maintained for now and migrated
+to the standalone client in the future.
+
+These bugs should be considered for the existing plugin.
+
+https://pagure.io/freeipa/issue/3096 - error when migrating unknown schema
+https://pagure.io/freeipa/issue/3100 - Check for userPassword in migration
+https://pagure.io/freeipa/issue/4738 - [RFE] ipa migrate-ds should provide option for creating UPG from posixGroup objectClass
+https://pagure.io/freeipa/issue/5020 - migrate-ds: does not show migrated users if an error happened during group migration
+https://pagure.io/freeipa/issue/5693 - Passwords become "expired" when migrating from directory server to IPA
+https://pagure.io/freeipa/issue/6105 - migrate-ds is not completely ignoring attributes.
+https://pagure.io/freeipa/issue/6360 - ipa migrate-ds does not rename uniquemember/member attributes properly
+https://pagure.io/freeipa/issue/6380 - ipa migrate-ds should print warning for referrals
+https://pagure.io/freeipa/issue/7368 - ipa migrate-ds converts groupofuniquenames objects to groupofnames, but leaves groupofuniquenames objectclass present
+https://pagure.io/freeipa/issue/7749 - `ipa migrate-ds` fails to migrate user and group data from directory server to IDM.
+
```

Comment:
Formatting: the above lines are concatenated into a single one. You can mark each one as a list element for a nicer output.

Reply:
Comments addressed.

I also renamed the design to ipa_to_ipa_migration.md to be a bit more specific.